### PR TITLE
add enroll app first version

### PR DESCRIPTION
**Pull request by Liang1XZhao**
- ehsm_srv add support for the RA_HANDSHAKE_MSG0 restful API
	this restful API will be exposed to public as:
	POST <ehsm_srv_address>/ehsm?Action=RA_HANDSHAKE_MSG0
	
- change the dockerfile
	add jsoncpp
	
- replaced Json process tool
	change to JSONCPP.
	
- Enroll App update content
	Added processing for msg0 and msg1.

test: passed the unittest

Signed-off-by: Liang1XZhao <Liang1XZhao@intel.com>

### CODE REVIEW

File: core/App/ehsm_napi.h [hunk 1]
```diff
@@ -50,112 +59,73 @@ static char* StringToChar(string str)
     return retChar;
 }
 
-typedef struct {
+struct RetJsonObj{
     const int CODE_SUCCESS = 200;
```
Comment (syan10) [inline core/App/ehsm_napi.h:63]: if you want to refine the struct, please submit another PR sperately. don't mix it in this patch.

File: core/App/ehsm_napi.h [hunk 2]
```diff
@@ -322,6 +292,12 @@ char* NAPI_Verify(const char* cmk_base64,
         const char* signature_base64);
 
 
+char* NAPI_RA_HANDSHAKE_MSG0(const char* request);
```
Comment (syan10) [inline core/App/ehsm_napi.h:295]: what's the request structure? please add comments to decript the details about the structure.

File: ehsm_kms_service/enroll_app/Makefile [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+#
+
+include ../../buildenv.mk
+
+App_Name := enroll
+
```
Comment (syan10) [inline ehsm_kms_service/enroll_app/Makefile:35]: name it to: ehsm-kms_enroll

File: ehsm_kms_service/enroll_app/main.cpp [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        for (j = 0; (j < i - 1); j++)
+            decode_str += char_array_3[j];
+    }
+
+    return decode_str;
+}
+
+static char* StringToChar(string str)
+{
+    char* returnChar = nullptr;
```
Comment (syan10) [inline ehsm_kms_service/enroll_app/main.cpp:166]: same here

File: ehsm_kms_service/enroll_app/main.cpp [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+// curl callback func
+size_t req_callback(void *ptr, size_t size, size_t nmem, void *stream)
+{
+	string *str = (string*)stream;
+	(*str).append((char*)ptr, size*nmem);
+	return size * nmem;
+}
+
+string post(string url, string content){
```
Comment (syan10) [inline ehsm_kms_service/enroll_app/main.cpp:204]: put the rest common functions into another file, please reference the base64.cpp/base64.h

File: docker/Dockerfile [hunk 6]
```diff
@@ -52,11 +52,19 @@ RUN apt-get update &&apt-get install -y \
...
 
 
-# Build App from source
+#  Install jsoncpp && Build App from source
 WORKDIR /home
-RUN git clone https://github.com/intel/ehsm.git \
-   && cd ehsm \
-   && make
+RUN git clone https://github.com/microsoft/vcpkg.git \
+    && git clone -b enroll_M0 https://github.com/Liang1XZhao/ehsm.git \
```
Comment (yang8621) [inline docker/Dockerfile:58]: https://github.com/Liang1XZhao/ehsm.git ?

File: ehsm_kms_service/ehsm_kms_server.js [hunk 7]
```diff
@@ -1,369 +1,199 @@
...
-const crypto = require('crypto')
-const app = express();
-const logger = require('./logs/logger');
-const ehsm_napi = require('./ehsm_napi');
-const { ehsm_kms_params, ehsm_keyspec_t, ehsm_keyorigin_t} = require('./ehsm_kms_params.js')
-const userInfo = require('./user_info')
-app.use(express.json());
 
-const PORT = process.argv.slice(2)[0] || 9000;
+const logger = require('./logger')
```
Comment (yang8621) [inline ehsm_kms_service/ehsm_kms_server.js:3]: Why commit this file? it's in #47 .

File: ehsm_kms_service/enroll_app/main.cpp [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+	Json::Value kmsConfigJson = readConfigJson();
+	string ehsm_kms_url = kmsConfigJson["ehsm_kms_url"].asString();
+	if(ehsm_kms_url.empty()){
+		printf("Error: ehsm_kms_url undefined.\n");
+        return ENL_CONFIG_INVALID;
+	}
+
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    string challenge = std::to_string(tv.tv_sec) + to_string(tv.tv_usec);
```
Comment (yang8621) [inline ehsm_kms_service/enroll_app/main.cpp:257]: aim to generate a random?

File: docker/Dockerfile [hunk 9]
```diff
@@ -52,11 +52,19 @@ RUN apt-get update &&apt-get install -y \
...
+    && cd vcpkg \
+    && ./bootstrap-vcpkg.sh \
+    && ./vcpkg integrate install \
+    && ./vcpkg install jsoncpp \
+    && cp ./installed/x64-linux/lib/libjsoncpp.a /usr/local/lib \
+    && cp -r ./installed/x64-linux/include/json /home/ehsm/include \
+    && cp -r /usr/include/x86_64-linux-gnu/curl /home/ehsm/include \
```
Comment (yang8621) [inline docker/Dockerfile:65]: can the below work?
apt install libjsoncpp-dev, libcurl4-openssl-dev or libcurl4-gnutls-dev

File: docker/Dockerfile (hunk 6 above)
Comment (Liang1XZhao) [reply docker/Dockerfile:58]: it will be fix

File: ehsm_kms_service/ehsm_kms_server.js (hunk 7 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/ehsm_kms_server.js:3]: The NAPI_RA_HANDSHAKE_MSG0 function is added on the basis of 47

File: docker/Dockerfile (hunk 9 above)
Comment (yang8621) [reply docker/Dockerfile:65]: Can you try this:
1. apt install libjsoncpp-dev
2. add #include <jsoncpp/json/json.h> to your file
3. g++ compile

File: core/App/ehsm_napi.h (hunk 1 above)
Comment (Liang1XZhao) [reply core/App/ehsm_napi.h:63]: ok

File: core/App/ehsm_napi.h (hunk 2 above)
Comment (Liang1XZhao) [reply core/App/ehsm_napi.h:295]: It will be fix

File: ehsm_kms_service/enroll_app/Makefile (hunk 3 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/enroll_app/Makefile:35]: It will be change

File: ehsm_kms_service/enroll_app/main.cpp (hunk 4 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/enroll_app/main.cpp:166]: It will be fix

File: ehsm_kms_service/enroll_app/main.cpp (hunk 5 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/enroll_app/main.cpp:204]: It will be fix

File: docker/Dockerfile [hunk 10]
```diff
@@ -52,11 +52,19 @@ RUN apt-get update &&apt-get install -y \
...
 WORKDIR /home
-RUN git clone https://github.com/intel/ehsm.git \
-   && cd ehsm \
-   && make
+RUN git clone https://github.com/microsoft/vcpkg.git \
+    && git clone -b enroll_M0 https://github.com/Liang1XZhao/ehsm.git \
+    && cd vcpkg \
+    && ./bootstrap-vcpkg.sh \
+    && ./vcpkg integrate install \
+    && ./vcpkg install jsoncpp \
```
[the remark this answers is not in the document]
Comment (Liang1XZhao) [reply docker/Dockerfile:62]: we need discuss.

File: ehsm_kms_service/enroll_app/main.cpp (hunk 8 above)
Comment (Liang1XZhao) [reply ehsm_kms_service/enroll_app/main.cpp:257]: yes,a timestamp

File: docker/Dockerfile (hunk 9 above)
Comment (Liang1XZhao) [reply docker/Dockerfile:65]: I will try it
